pkg/providers/tinkerbell/reconciler/reconciler.go

```diff
@@ -111,17 +130,97 @@ func (r *Reconciler) ValidateClusterSpec(ctx context.Context, log logr.Logger, t
+
+	w, err := tinkerbell.WorkersSpec(ctx, log, clientutil.NewKubeClient(r.client), spec)
+	if err != nil {
+		return controller.Result{}, errors.Wrap(err, "generating workers spec")
+	}
+	tinkerbellScope.Workers = w
 
-	return clusters.ReconcileControlPlane(ctx, r.client, toClientControlPlane(cp))
+	return controller.Result{}, nil
+}
+
+// DetectOperation detects change type.
+func (r *Reconciler) DetectOperation(ctx context.Context, log logr.Logger, tinkerbellScope *Scope) (controller.Result, error) {
+	log = log.WithValues("phase", "DetectOperation")
+	log.Info("Detecting operation type")
+
+	currentKCP, err := controller.GetKubeadmControlPlane(ctx, r.client, tinkerbellScope.ClusterSpec.Cluster)
+	if err != nil {
+		return controller.Result{}, err
+	}
+	if currentKCP == nil {
+		tinkerbellScope.ClusterChange = ReconcileNewCluster
+		return controller.Result{}, nil
+	}
+
+	tinkCurrentKCP := tinkerbell.KubeadmControlPlane{KubeadmControlPlane: currentKCP}
+	tinkDesiredKCP := tinkerbell.KubeadmControlPlane{KubeadmControlPlane: tinkerbellScope.ControlPlane.KubeadmControlPlane}
+	wantVersionChange := tinkerbell.HasKubernetesVersionChange(&tinkCurrentKCP, &tinkDesiredKCP)
+	cpWantScaleChange := tinkerbell.ReplicasDiff(&tinkCurrentKCP, &tinkDesiredKCP)
+	workerWantScaleChange := r.workerReplicasDiff(ctx, tinkerbellScope)
+
+	if wantVersionChange {
+		if cpWantScaleChange != 0 || workerWantScaleChange {
+			return controller.Result{}, errors.Errorf("cannot perform scale up or down during k8s version change")
+		}
+		tinkerbellScope.ClusterChange = NeedNewNode
+	} else if cpWantScaleChange != 0 || workerWantScaleChange {
+		tinkerbellScope.ClusterChange = NeedMoreOrLessNode
+	} else {
+		return controller.Result{}, errors.Errorf("cannot detect operation type")
+	}
+	return controller.Result{}, nil
+}
+
+func (r *Reconciler) workerReplicasDiff(ctx context.Context, tinkerbellScope *Scope) bool {
+	workerWantScaleChange := false
+	for _, wnc := range tinkerbellScope.ClusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations {
+		md := &clusterv1.MachineDeployment{}
+		key := types.NamespacedName{Namespace: constants.EksaSystemNamespace, Name: wnc.Name}
+		err := r.client.Get(ctx, key, md)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				workerWantScaleChange = true
+				break
+			}
+		}
+		if int(*md.Spec.Replicas) != *wnc.Count {
+			workerWantScaleChange = true
+			break
+		}
+	}
+	return workerWantScaleChange
+}
+
+// OmitMachineTemplate omits control plane and worker machine template on scaling update.
+func (r *Reconciler) OmitMachineTemplate(ctx context.Context, log logr.Logger, tinkerbellScope *Scope) (controller.Result, error) {
+	log = log.WithValues("phase", "OmitMachineTemplate")
+	log.Info("Omit machine template on scaling")
+	if tinkerbellScope.ClusterChange == NeedMoreOrLessNode {
+		tinkerbell.OmitTinkerbellCPMachineTemplate(tinkerbellScope.ControlPlane)
+		tinkerbell.OmitTinkerbellWorkersMachineTemplate(tinkerbellScope.Workers)
```

Comment:
can't we just have this as part of the capi spec generation methods?

Reply:
machine template would be needed in hardware validation. @chrisdoherty4  @raymond-zhang00

Reply:
I think I was mistaken, the omits only affect the desired CAPI objects rather than the current CAPI objects. The `AssertionsForScaleUpDown` just need the current machine templates in order to retrieve the hardware selector. It gets desired state hardware selector via the Tinkerbell ClusterSpec. 

We should actually be okay to remove this before the assertion.

Reply:
I am kinda confused here. Technically we dont need to omit right? Isn't there a logic in the template generation to just fetch reuse the same machine template on the cluster already?

I believe it was here https://github.com/aws/eks-anywhere/blob/main/pkg/clusterapi/controlplane.go#L51

And as of the comparator we just wanted to add a condition based on scale up/rolling? 
https://github.com/aws/eks-anywhere/blob/main/pkg/providers/tinkerbell/machinetemplate.go#L25

Reply:
What about still omitting it in a separate phase, but before assertion?